File: C++/DataStructure&oj/mergeSort.cpp
#include<cstdio>
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
using namespace std;

const int maxn=100000;
int a[maxn]={0};
int aux[maxn]={0};

void Merge(int *a,int be,int mid,int en);
void mergeSort(int *a,int be,int en)
{
    if(be>=en)
        return;
    int mid=be+(en-be)/2;
    mergeSort(a,be,mid);
    mergeSort(a,mid+1,en);
    Merge(a,be,mid,en);
}

void Merge(int *a,int be,int mid,int en)
{
    for(int i=be;i<en+1;++i)
        aux[i]=a[i];
    int i=be;
    int j=mid+1;
    int k=be;
    while(k<en+1)
    {
        if(i>mid)
        {
            a[k++]=aux[j++];
        }
        else if(j>en)
        {
            a[k++]=aux[i++];
        }
        else if(aux[i]<aux[j])
        {
            a[k++]=aux[i++];
        }
        else
        {
            a[k++]=aux[j++];
        }
    }
}

int main()
{
    int n;
    scanf("%d",&n);
    while(n--)
    {
        int m;
        scanf("%d",&m);
        for(int i=0;i<m;++i)
        {
            scanf("%d",&a[i]);
        }
        mergeSort(a,0,m-1);
        for(int i=0;i<m;++i)
        {
            printf("%d\n",a[i]);
        }
    }
}

File: Java/Algorithm/MergeSort.java
public class MergeSort
{
    private static Comparable[] aux;

    public static void sort(Comparable[] a)
    {
        aux= new Comparable[a.length];
        sort(a,0,a.length-1);
    }

    private static void sort(Comparable[] a,int lo,int hi)
    {
        if(lo>=hi)
        {
            return;
        }
        int mid =lo+(hi-lo)/2;
        sort(a,lo,mid);
        sort(a,mid+1,hi);
        merge(a,lo,mid,hi);
    }

    private static void merge(Comparable[] a,int lo,int mid,int hi)
    {
        int i=lo;
        int j=mid+1;
        for(int s=lo;s<hi+1;++s)
        {
            aux[s]=a[s];
        }
        for(int k=i;k<j;++k)
        {
            if(i>mid)
            {
                a[k]=aux[j++];
            }
            else if(j>hi)
            {
                a[k]=aux[i++];
            }
            else if(less(aux[j],aux[i]))
            {
                a[k]=aux[j++];
            }
            else
            {
                a[k]=aux[i++];
            }
        }
    }

    private static boolean less(Comparable p,Comparable q)
    {
        return p.compareTo(q)<0;
    }

    private static void exch(Comparable[] a,int i,int j)
    {
        Comparable temp=a[i];
        a[i]=a[j];
        a[j]=temp;
    }

    public static void show(Comparable[] a)
    {
        for(Comparable i:a)
        {
            System.out.print(i+" ");
        }
        System.out.println();
    }

    public static boolean isSorted(Comparable[] a)
    {
        for(int i=0;i<a.length-1;++i)
        {
            if(less(a[i+1],a[i]))
            {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args)
    {
        Integer[] a=new Integer[20];
        for(int i=0;i<a.length;++i)
        {
            a[i]=(int)(Math.random()*100);
        }
        MergeSort.show(a);
        MergeSort.sort(a);
        MergeSort.show(a);
    }
}
